builder/src/lib.rs

```diff
@@ -1,10 +1,247 @@
 extern crate proc_macro;
+use std::collections::{HashMap, HashSet};
 
+use quote::{quote};
 use proc_macro::TokenStream;
+use proc_macro2::{TokenStream as TokenStream2};
+use syn::{DeriveInput, parse_macro_input, Data, Fields, Ident};
 
-#[proc_macro_derive(Builder)]
+#[proc_macro_derive(Builder, attributes(builder))]
 pub fn derive(input: TokenStream) -> TokenStream {
-    let _ = input;
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+    let buildername = Ident::new(&format!("{}Builder", name), input.ident.span().clone());
 
-    unimplemented!()
+    let (fields, fields_init, setters, build) = match input.data {
+        Data::Struct(ref data) => {
+            match data.fields {
+                Fields::Named(ref fields) => {
+                    let builders = match get_fields_builders(&fields) {
+                        Some(b) => b,
+                        None => return TokenStream::from(quote! {
+                            compile_error!("missing or invalid `each` parameter in builder");
+                        }),
+                    };
+                    let optional_fields = get_optional_fields(&fields);
+                    (
+                        expand_field_definitions(&fields, &optional_fields, &builders),
+                        expand_field_initializers(&fields, &builders),
+                        expand_field_setters(&fields, &optional_fields, &builders),
+                        expand_build(&name, &fields, &optional_fields, &builders),
+                    )
+                }
+                _ => { (quote!{}, quote!{}, quote!{}, quote!{}) },
+            }
+        },
+        _ => unimplemented!(),
+    };
+
+    let expanded = quote! {
+        impl #name {
+            pub fn builder() -> #buildername {
+                #buildername { #fields_init }
+            }
+        }
+
+        pub struct #buildername {
+            #fields
+        }
+
+        impl #buildername {
+            #setters
+
+            pub fn build(&mut self) -> std::result::Result<#name, std::boxed::Box<dyn std::error::Error>> {
+                #build
+            }
+        }
+    };
+    TokenStream::from(expanded)
+}
+
+fn expand_field_definitions(fields: &syn::FieldsNamed, optional_fields: &HashSet<String>, builders: &HashMap<String, String>) -> TokenStream2 {
```

Comment:
Would it be possible to have this set and map as HashSet\<Ident\> and HashMap\<Ident, Ident\> everywhere? I think that would cut down on the code here going back and forth between Ident and String and end up being more readable.